src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs

```diff
@@ -1022,35 +1031,36 @@ public string ToStringClass()
             // This is important because if the last range ends in LastChar, we won't append
             // LastChar to the list.
             int rangeLen = _rangelist.Count * 2;
-            StringBuilder sb = StringBuilderCache.Acquire(rangeLen + _categories.Length + 3);
+            int strGuessCount = rangeLen + _categories.Length + 3;
 
-            int flags;
-            if (_negate)
-                flags = 1;
-            else
-                flags = 0;
+            Span<char> buffer = strGuessCount <= 256 ? stackalloc char[256] : null;
+            ValueStringBuilder vsb = buffer != null ?
+                new ValueStringBuilder(buffer) :
+                new ValueStringBuilder(strGuessCount);
+
+            int flags = _negate ? 1 : 0;
 
-            sb.Append((char)flags);
-            sb.Append((char)rangeLen);
-            sb.Append((char)_categories.Length);
+            vsb.Append((char)flags);
+            vsb.Append((char)rangeLen);
+            vsb.Append((char)_categories.Length);
 
             for (int i = 0; i < _rangelist.Count; i++)
             {
                 SingleRange currentRange = _rangelist[i];
-                sb.Append(currentRange.First);
+                vsb.Append(currentRange.First);
 
                 if (currentRange.Last != LastChar)
-                    sb.Append((char)(currentRange.Last + 1));
+                    vsb.Append((char)(currentRange.Last + 1));
             }
 
-            sb[SETLENGTH] = (char)(sb.Length - SETSTART);
+            vsb[SETLENGTH] = (char)(vsb.Length - SETSTART);
 
-            sb.Append(_categories);
+            vsb.Append(_categories.ToString());
 
             if (_subtractor != null)
-                sb.Append(_subtractor.ToStringClass());
+                vsb.Append(_subtractor.ToStringClass());
```

Comment:
ToStringClass is implemented now to create and write to a ValueStringBuilder.  So this is creating a new ValueStringBuilder, writing to it, creating a string from it, and then appending that string to this ValueStringBuilder.  You could instead slightly refactor ToStringClass to accept the VSB to write into, and then just pass in this existing vsb.

Reply:
I spent some time now trying to implement your recommendations but the ToStringClass method is used in couple of places where it's hard to introduce the Span & ValueStringBuilder locals without risking performance degradation in other branches. I think I will leave it as it is for now and tackle it as part of https://github.com/dotnet/corefx/issues/30507

Reply:
that's the specific caller I'm talking about: http://source.dot.net/#System.Text.RegularExpressions/System/Text/RegularExpressions/RegexParser.cs,1067